### ISSUE TITLE
Incorrect return type

### ISSUE THREAD

**Issue by tdbgamer** [redox-os/relibc#89]
https://github.com/redox-os/relibc/blob/d75535974ae37e7d698c01c46584705b790e0bc9/src/string/src/lib.rs#L277

I believe this function is supposed to return `size_t`. Actually it looks like all the functions in that file that return `c_ulong` should return `size_t`. I modified the type and the tests passed fine for me. Should I open a PR?

**Comment by dlrobertson** [comment]
> Should I open a PR?

That would be great!

**Comment by tdbgamer** [comment]
I just implemented `strtok` as well but it's using those functions that I changed in this PR. So once this is merged I'll make another PR for `strtok`.

**Comment by ids1024** [comment]
I wonder if there's some way we could automatically test if the generated headers match C/POSIX standards...

**Comment by xTibor** [comment]
How about some tests something like this? I've seen makefiles using similar techniques. However no idea how robust is this. https://gist.github.com/xTibor/c0cd9fa9d05c042c4cf913e04988a6cf

**Comment by ids1024** [comment]
Something like that might work; the alternative would be to write a program that parses the C headers and compares the function signatures (with a good C parsing library, that might work alright, and we could configure the behavior more).

Is there anywhere we can get .h files for these standards? Other than parsing html files like http://pubs.opengroup.org/onlinepubs/9699919799/basedefs/string.h.html...

**Comment by ids1024** [comment]
> Other than parsing html files...

...which would look like this, but still has some issues: https://gist.github.com/1530bca58dace62971dbbf8270f36c6d

**Comment by ids1024** [comment]
I've hacked something together. It's incomplete, and the script only runs on stdlib.h for now.

https://gist.github.com/ids1024/9722da9fc3f2beb621f4c9527cf3db70

```
Argument 1 of memccpy() is type 'const const void *' on Redox, 'const void * restrict' in SUS
Argument 3 of memccpy() is type 'uintptr_t' on Redox, 'size_t' in SUS
Argument 2 of memchr() is type 'uintptr_t' on Redox, 'size_t' in SUS
Argument 1 of strcat() is type 'const const char *' on Redox, 'const char * restrict' in SUS
Argument 1 of strcmp() is type 'const const char *' on Redox, 'const char *' in SUS
Argument 1 of strcoll() is type 'const const char *' on Redox, 'const char *' in SUS
Argument 1 of strcpy() is type 'const const char *' on Redox, 'const char * restrict' in SUS
Argument 1 of strcspn() is type 'const const char *' on Redox, 'const char *' in SUS
'strerror()' definition matches
Argument 1 of strncat() is type 'const const char *' on Redox, 'const char * restrict' in SUS
Argument 2 of strncat() is type 'uintptr_t' on Redox, 'size_t' in SUS
Argument 1 of strncmp() is type 'const const char *' on Redox, 'const char *' in SUS
Argument 2 of strncmp() is type 'uintptr_t' on Redox, 'size_t' in SUS
Argument 1 of strncpy() is type 'const const char *' on Redox, 'const char * restrict' in SUS
Argument 2 of strncpy() is type 'uintptr_t' on Redox, 'size_t' in SUS
Argument 1 of strndup() is type 'uintptr_t' on Redox, 'size_t' in SUS
Argument 1 of strnlen() is type 'uintptr_t' on Redox, 'size_t' in SUS
Argument 1 of strpbrk() is type 'const const char *' on Redox, 'const char *' in SUS
Argument 1 of strspn() is type 'const const char *' on Redox, 'const char *' in SUS
Argument 1 of strstr() is type 'const const char *' on Redox, 'const char *' in SUS
Argument 1 of strtok() is type 'const const char *' on Redox, 'const char * restrict' in SUS
Argument 1 of strtok_r() is type 'const const char *' on Redox, 'const char * restrict' in SUS
Argument 2 of strtok_r() is type 'char * *' on Redox, 'char * * restrict' in SUS
Argument 1 of strxfrm() is type 'const const char *' on Redox, 'const char * restrict' in SUS
Argument 2 of strxfrm() is type 'uintptr_t' on Redox, 'size_t' in SUS
```

It works well enough to provide somewhat interesting output; I think this could be improved, and be a reasonable solution.

**Comment by tdbgamer** [comment]
Maybe the redox types should be named the same way as in the header files so it'll be easier to make some assertions in that script?

**Comment by dlrobertson** [comment]
Core issue fixed by https://github.com/redox-os/relibc/pull/91. Finding a way to automagically check the function signature was split out into #111.
